src/borg/archive.py

```diff
@@ -719,24 +719,39 @@ def process_symlink(self, path, st):
         return 's'  # symlink
 
     def chunk_file(self, item, cache, stats, fd, fh=-1, **chunk_kw):
-        checkpoint_number = 1
+        def write_part(item, from_chunk, number):
+            item = Item(internal_dict=item.as_dict())
+            length = len(item.chunks)
+            # the item should only have the *additional* chunks we processed after the last partial item:
+            item.chunks = item.chunks[from_chunk:]
+            item.path += '.checkpoint_%d' % number
+            item.checkpoint = number
+            number += 1
+            self.add_item(item, show_progress=False)
+            self.write_checkpoint()
+            # we have saved the checkpoint file, but we will reference the same
+            # chunks also from the final, complete file:
+            for chunk in item.chunks:
+                cache.chunk_incref(chunk.id, stats)
```

Comment:
hmm, i guess this is wrong since the "parts" change.

before, the duplicate references of the leading chunks happened with every checkpoint commit, as the beginnings of the checkpoint files were overlapping, like abc, abcdef, abcdefghi.
but since working with parts, we do not have the overlap any more while we are doing part files:
abc, def, ghi.
we need to duplicate the references after saving the last part, when we save the final (non-part) full file.